include/tc/core/cuda/cuda_mapping_options.h

```diff
@@ -0,0 +1,248 @@
+  inline explicit CudaMappingOptions(const std::string& str);
+
+  std::string toProtobufSerializedString() const {
+    return ownedProto_.SerializeAsString();
+  }
+
+  /**
+   * @name Chainable Modifiers specific to CudaMappingOptions
+   * See protobuf for documentation on each option.
+   * @{
+   */
+  /// Set mappings
+  ///@{
+  inline CudaMappingOptions& mapToThreads(
+      std::initializer_list<uint64_t> threads);
+  inline CudaMappingOptions& mapToThreads(
+      uint64_t x,
+      uint64_t y = CudaDimView::defaultDim,
+      uint64_t z = CudaDimView::defaultDim);
+  inline CudaMappingOptions& mapToThreads(const std::vector<uint64_t>& threads);
+  CudaMappingOptions& mapToThreads(const std::string& commaSeparatedSizes);
+
+  inline CudaMappingOptions& mapToBlocks(
+      std::initializer_list<uint64_t> blocks);
+  inline CudaMappingOptions& mapToBlocks(
+      uint64_t x,
+      uint64_t y = CudaDimView::defaultDim,
+      uint64_t z = CudaDimView::defaultDim);
+  inline CudaMappingOptions& mapToBlocks(const std::vector<uint64_t>& blocks);
+  CudaMappingOptions& mapToBlocks(const std::string& commaSeparatedSizes);
+  ///@}
+
+  /// Set mappings
+  inline CudaMappingOptions& genericMappingOptions(
+      const MappingOptions& options);
+  inline CudaMappingOptions& useSharedMemory(bool b);
+  inline CudaMappingOptions& usePrivateMemory(bool b);
+  inline CudaMappingOptions& maxSharedMemory(uint64_t size);
+  inline CudaMappingOptions& unrollCopyShared(bool b);
+  ///@}
+
+  /// Static constructors for predefined strategies.
+  ///@{
+  static CudaMappingOptions makeNaiveCudaMappingOptions();
+  static CudaMappingOptions makeSingleThreadCudaMappingOptions();
+  static CudaMappingOptions makePointwiseCudaMappingOptions();
+  static CudaMappingOptions makeMlpCudaMappingOptions();
+  static CudaMappingOptions makeConvolutionCudaMappingOptions();
+  static CudaMappingOptions makeGroupConvolutionCudaMappingOptions();
+  ///@}
+
+  const CudaMappingOptionsProto& proto() const {
+    return ownedProto_;
+  }
+
+#define FORWARD_FUN(FUN_NAME)                         \
+  template <typename... Args>                         \
+  inline CudaMappingOptions& FUN_NAME(Args... args) { \
+    genericMappingOptionsView.FUN_NAME(args...);      \
+    return *this;                                     \
+  }
+
+  FORWARD_FUN(tile);
+  FORWARD_FUN(unroll);
+  FORWARD_FUN(fixParametersBeforeScheduling);
+  FORWARD_FUN(tileImperfectlyNested);
+  FORWARD_FUN(matchLibraryCalls);
+  FORWARD_FUN(scheduleFusionStrategy);
+  FORWARD_FUN(outerScheduleFusionStrategy);
+  FORWARD_FUN(outerScheduleAllowSkewing);
+  FORWARD_FUN(outerSchedulePositiveOrthant);
+  FORWARD_FUN(intraTileScheduleFusionStrategy);
+  FORWARD_FUN(intraTileScheduleAllowSkewing);
+  FORWARD_FUN(intraTileSchedulePositiveOrthant);
+
+#undef FORWARD_FUN
+
+ private:
+  CudaMappingOptionsProto ownedProto_;
+
+ public:
+  MappingOptionsView genericMappingOptionsView;
```

Comment:
Unclear, so far my experience has been that having a concrete object inherit from a view creates all sorts of hellish allocation order issues and tricky segfaults.
This would be solved more easily if proto supported inheritance but unfortunately it does not.